integrations/digital/outlook.js

```diff
@@ -0,0 +1,143 @@
+// function getActivityFromEmail(message) {
+//   return {
+//     id: message.id,
+//     activityType: ACTIVITY_TYPE_DIGITAL,
+//     lineItems: [{ identifier: DIGITAL_CATEGORY_EMAIL, unit: UNIT_ITEM, value: 1 }],
+//     locationLabel: 'Outlook',
+//     label: 'Email received',
+//     carrier: 'Microsoft',
+//     datetime: moment.parseZone(message.lastModifiedDateTime).toDate(),
+//   };
+// }
+async function fetchRemainingEmailActivities(client, nextLink) {
+  const resultActivities = [];
+  const messages = await client.api(nextLink).get();
+  let resultDeltaLink = messages[odataDeltaLinkPropertyName];
+  if (messages[odataNextLinkPropertyName]) {
+    const { activities, deltaLink } = await fetchRemainingEmailActivities(
+      client,
+      messages[odataNextLinkPropertyName]
+    );
+    resultActivities.push(...activities);
+    resultDeltaLink = deltaLink;
+  }
+  resultActivities.push(
+    ...flatten(
+      messages.value.map(x =>
+        getActivitiesFromEmail(
+          x.subject,
+          x.from && x.from.emailAddress && x.from.emailAddress.address,
+          x.body && x.body.content,
+          new Date(x.lastModifiedDateTime)
+        )
+      )
+    )
+  );
+  // resultActivities.push(...messages.value.map(x => getActivityFromEmail(x)));
+  // TODO: uncomment line above and corresponding function when digital footprint is enabled
+  return { activities: resultActivities, deltaLink: resultDeltaLink };
+}
+async function fetchEmail(client, previousDeltaLink, folderId) {
+  return {
+    ...(await fetchRemainingEmailActivities(
+      client,
+      previousDeltaLink === undefined
+        ? `me/mailFolders/${folderId}/messages/delta?$select=id,lastModifiedDateTime,body,subject,from`
+        : previousDeltaLink
+    )),
+    folderId,
+  };
+}
+
+async function fetchEmailFolders(client, nextLink) {
+  const result = [];
+  const response = await client
+    .api(nextLink === undefined ? `me/mailFolders?$select=id` : nextLink)
+    .get();
+  result.push(...response.value.map(x => x.id));
+  if (response[odataNextLinkPropertyName] !== undefined) {
+    result.push(...(await fetchEmailFolders(client, response[odataNextLinkPropertyName])));
+  }
+  return result;
+}
+
+async function connect(requestLogin, requestWebView) {
+  const state = await manager.authorize(requestWebView);
+  return state;
+}
+
+async function disconnect() {
+  return {};
+}
+
+async function collect(state = {}, logger) {
+  manager.setState(state);
+
+  logger.logDebug(`Initiating collect() for Outlook`);
+
+  const client = Client.init({
+    authProvider: done => {
+      done(undefined, state.accessToken);
+    },
+  });
+
+  if (!state.emailFolders) {
+    state.emailFolders = (await fetchEmailFolders(client)).map(x => {
+      return { folderId: x, deltaLink: undefined };
+    });
+    logger.logDebug(`Found ${state.emailFolders.length} email folders for Outlook`);
+  }
+
+  const allResults = await Promise.all(
```

Comment:
Can this potentially be a slow and network intensive operation?
Would it make sense to keep track of progress to make sure we don't reprocess emails that have already been processed? Maybe that becomes a bit more complicated as new email parsers get added though.

Reply:
I'm not expecting inboxes to have more than a dozen folders in most cases. Using promise all is probably the best way to make sure the runtime handles the work the best it can depending on things like event loop, pool sizes and more. https://itnext.io/node-js-handling-asynchronous-operations-in-parallel-69679dfae3fc

The only thing we could potentially run into at this point depending on the runtime is the limit of 4 parallel requests per mailbox. But you'd see 429's in your telemetry and that's something we could easily address by batching by groups of 4 promises (see 3rd example in article).

I'm using Delta links in this case, not the list API. Delta links give you the changes and a token that is a watermark so that whenever you call it back with the token it will only give you changes since last call.
https://docs.microsoft.com/en-us/graph/api/message-delta?view=graph-rest-1.0&tabs=http

Reply:
But having a way to give visual indications to the user that progress is being made would be great. Something like an infinite progress bar with a message like 2000 emails analyzed, 30 activities found...
However the prototype for integrations doesn't seem to support it at this time.

Reply:
Ok noted - we should be fine for now then.